Mathlib/NumberTheory/NumberField/FinitePlaces.lean

```diff
@@ -0,0 +1,242 @@
+    ∀ v : HeightOneSpectrum (𝓞 K), ‖embedding v x‖ = r := Set.forall_subtype_range_iff
+
+theorem le_iff_le (x : K) (r : ℝ) : (∀ w : FinitePlace K, w x ≤ r) ↔
+    ∀ v : HeightOneSpectrum (𝓞 K), ‖embedding v x‖ ≤ r := Set.forall_subtype_range_iff
+
+theorem pos_iff {w : FinitePlace K} {x : K} : 0 < w x ↔ x ≠ 0 := AbsoluteValue.pos_iff w.1
+
+@[simp]
+theorem mk_eq_iff {v₁ v₂ : HeightOneSpectrum (𝓞 K)} : mk v₁ = mk v₂ ↔ v₁ = v₂ := by
+  refine ⟨?_, fun a ↦ by rw [a]⟩
+  contrapose!
+  intro h
+  rw [DFunLike.ne_iff]
+  have : ∃ x : 𝓞 K, x ∈ v₁.asIdeal ∧ x ∉ v₂.asIdeal := by
+    by_contra!
+    apply h
+    exact HeightOneSpectrum.ext_iff.mpr (IsMaximal.eq_of_le (isMaximal v₁) IsPrime.ne_top' this)
+  rcases this with ⟨x, hx1, hx2⟩
+  use x
+  simp only [apply]
+  rw [← norm_lt_one_iff_mem ] at hx1
+  rw [← norm_eq_one_iff_not_mem] at hx2
+  linarith
+
+theorem max_ideal_mk (v : HeightOneSpectrum (𝓞 K)) : maximal_ideal (mk v) = v := by
+  rw [← mk_eq_iff, mk_max_ideal]
+
+theorem mulSupport_Finite_int {x : 𝓞 K} (h_x_nezero : x ≠ 0) :
+    (Function.mulSupport fun w : FinitePlace K => w x).Finite := by
+  have (w : FinitePlace K) : w x ≠ 1 ↔ w x < 1 := by
+    have := norm_le_one w.maximal_ideal x
+    rw [norm_embedding_eq] at this
+    exact Decidable.ne_iff_lt_iff_le.mpr this
+  simp_rw [Function.mulSupport, this, ← norm_embedding_eq, norm_lt_one_iff_mem,
+    ← Ideal.dvd_span_singleton]
+  have h : {v : HeightOneSpectrum (𝓞 K) | v.asIdeal ∣ span {x}}.Finite := by
+    apply Ideal.finite_factors
+    simp only [Submodule.zero_eq_bot, ne_eq, span_singleton_eq_bot, h_x_nezero, not_false_eq_true]
+  have h_inj : Set.InjOn (fun w : FinitePlace K ↦ maximal_ideal w)
+      {x_1 | x_1.maximal_ideal.asIdeal ∣ span {x}} := by
+    apply Function.Injective.injOn
+    intro w₁ w₂ h
+    rw [← mk_max_ideal w₁, ← mk_max_ideal w₂]
+    exact congrArg mk h
+  apply Set.Finite.of_finite_image _ h_inj
+  have h_subs : ((fun w : FinitePlace K ↦ w.maximal_ideal) ''
+      {x_1 : FinitePlace K | x_1.maximal_ideal.asIdeal ∣ span {x}}) ⊆
+      {v : HeightOneSpectrum (𝓞 K) | v.asIdeal ∣ span {x}} := by
+    simp only [dvd_span_singleton, Set.image_subset_iff, Set.preimage_setOf_eq, subset_refl]
+  exact Set.Finite.subset h h_subs
+
+theorem mulSupport_Finite {x : K} (h_x_nezero : x ≠ 0) :
+    (Function.mulSupport fun w : FinitePlace K => w x).Finite := by
+  rcases @IsFractionRing.div_surjective (𝓞 K) _ K _ _ _ x with ⟨a, b, hb, hfrac⟩
+  subst hfrac
+  simp_all only [ne_eq, div_eq_zero_iff, NoZeroSMulDivisors.algebraMap_eq_zero_iff, not_or,
+    map_div₀]
+  obtain ⟨ha, hb⟩ := h_x_nezero
+  rw [← RingOfIntegers.coe_eq_algebraMap, ← RingOfIntegers.coe_eq_algebraMap]
+  apply Set.Finite.subset (Set.Finite.union (mulSupport_Finite_int ha) (mulSupport_Finite_int hb))
+  intro w
+  contrapose!
+  intro a_1
+  simp_all only [Set.mem_union, Function.mem_mulSupport, ne_eq, not_or, Decidable.not_not,
+    one_ne_zero, not_false_eq_true, div_self, not_true_eq_false]
```

Comment:
```suggestion
  simp only [Function.mem_mulSupport, ne_eq, Set.mem_union]
  contrapose!
  simp +contextual only [ne_eq, one_ne_zero, not_false_eq_true, div_self, implies_true]
```
The `apply` and `intro` lines can be combined (again, a matter of taste):
```lean
  refine ((mulSupport_Finite_int ha).union (mulSupport_Finite_int hb)).subset fun w ↦ ?_
```
This can be simplified a bit.